### ISSUE TITLE
Feature Request: Support for EC2 Capacity Reservations

### ISSUE THREAD

**Issue by gazoakley** [terraform-providers/terraform-provider-aws#6289]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS have announced a new Capacity Reservation feature to reserve capacity in AZs without making long term commitments. Reservations can be cancelled at any time, and instances can be attributed to specific capacity reservations:

* https://aws.amazon.com/about-aws/whats-new/2018/10/Amazon-EC2-now-offers-On-Demand-Capacity-Reservations/
* https://aws.amazon.com/ec2/faqs/#capacityreservations

### New or Affected Resource(s)

* `aws_ec2_capacity_reservation`
* `aws_launch_template`
* `aws_instance`

Instances can be also associated with a capacity reservation after being created, but they must be stopped to make the change. For now, changing the reservation on an `aws_instance` should probably be a `force_new`.

### Potential Terraform Configuration

```hcl
# New Resource, not implemented and details may change during implementation
resource "aws_ec2_capacity_reservation" "default" {
  instance_type     = "t2.micro"
  instance_platform = "Linux/UNIX"
  availability_zone = "eu-west-1a"
  instance_count    = 1
}
```

### References

* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateCapacityReservation.html
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CancelCapacityReservation.html
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyCapacityReservation.html
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyInstanceCapacityReservationAttributes.html

**Comment by bflad** [comment]
Thanks to @gazoakley this was quickly implemented and will release with version 1.42.0 of the AWS provider very shortly. 🎉

**Comment by bflad** [comment]
Ah, FYI this issue was closed due to the new `aws_ec2_capacity_reservation` resource. For implementing the capacity reservation support within other resources (e.g. `aws_instance`), I recommend creating other feature request issues so they can be individually tracked as they will likely take much longer to get in.

**Comment by bflad** [comment]
This has been released in [version 1.42.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1420-october-31-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
